compiler/ir/backend.common/src/org/jetbrains/kotlin/ir/util/IrTypeUtils.kt

```diff
@@ -89,3 +92,38 @@ fun IrType.getPrimitiveArrayElementType() = (this as? IrSimpleType)?.let {
 
 fun IrType.isNonPrimitiveArray() =
     (this.isArray() || this.isNullableArray()) && !this.isPrimitiveArray()
+
+fun IrType.eraseTypeParameters() = when (this) {
+    is IrErrorType -> this
+    is IrSimpleType -> {
+        val owner = classifier.owner
+        when (owner) {
+            is IrClass -> this
+            is IrTypeParameter -> {
+                val upperBound = owner.upperBoundClass()
+                IrSimpleTypeImpl(
+                    upperBound.symbol,
+                    hasQuestionMark,
+                    List(upperBound.typeParameters.size) { IrStarProjectionImpl },
+                    owner.annotations
+                )
+            }
+            else -> error("Unknown IrSimpleType classifier kind: $owner")
+        }
+    }
+    else -> error("Unknown IrType kind: $this")
+}
+
+fun IrTypeParameter.upperBoundClass(): IrClass {
+    // Look into super types for possible upperbound
+    val simpleSuperClassifiers = superTypes.asSequence().filterIsInstance<IrSimpleType>().map { it.classifier }
+    return simpleSuperClassifiers
+        .filterIsInstance<IrClassSymbol>()
+        .let {
+            // First try to find a non-interface class in super types, return it as upperbound.
+            it.firstOrNull { !it.owner.isInterface } ?: it.firstOrNull()
```

Comment:
1) Please don't use nested `it`s, declare one of them with a name instead
2) Take a look at [`representativeUpperBound`](https://github.com/JetBrains/kotlin/blob/373424f4304784cfda7fd7219838dbdd728bed54/core/descriptors/src/org/jetbrains/kotlin/types/TypeUtils.kt#L236) which implements similar behavior in compiler frontend. It also checks if the upper bound is an annotation (which is technically an interface on JVM). Also, consider even extracting this part of the function to a new function `IrTypeParameter.representativeUpperBound: IrType` just to be clear that this is exactly the same behavior, and then maybe moving the rest (possibly renaming to `representativeUpperBoundClass`) to JVM-specific utils

Reply:
UPD: I've noticed this code was already there in BridgeLowering before your changes. In that case, feel free to leave everything as is to avoid breaking something else. However, I feel very nervous about exposing this functionality in `backend.common`, I'd leave it in `backend.jvm` for now

Reply:
Makes sense, moved to backend.jvm